### DISCUSSION TITLE
Converting 3 bytes from i2c sensor to float

### DISCUSSION THREAD

**Discussion by labodezao** [micropython/micropython#10564] [STM32 / Pyboard] [answered]
Hi Here, I use this senos with i2c : https://cfsensor.com/product/piezoresistive-pressure-sensor-xgzp6847d/
and the micropython code gives weird results : 

     
        self.buf = self.bus.readfrom_mem(0x6d, 0x06, 5)
        #take the first 2 bytes of measurements
        print(self.buf[0],self.buf[1],self.buf[2])
        pressure_adc = int(self.buf[0]) * 65536 + int(self.buf[1]) * 256 + int(self.buf[2]);
        #Compute the value of pressure converted by ADC
       
        pressure_value=ustruct.unpack(">h",bytearray([self.buf[0],self.buf[1],self.buf[2]]))[0]
        self.d_pres= pressure_value /self.scale_press
        t_value=ustruct.unpack(">h",self.buf[3:5] )[0]
        self.temp= t_value

I use struct packahe for convert my signed int to float, and the result is not really fitting with reality, but slightly, but when I do it manually like this : 

    
        if (pressure_adc > 8388608):
            pressure = (pressure_adc - 16777216) / self.scale_press
        else:
            pressure = pressure_adc / self.scale_press
            
        print('adc',pressure_adc)

I get a tottaly différent result from ustruct method, and I don't understand why ...
The results is someting like 4 with ustruct and 12750 with manual method ...

**Comment by robert-hh** [comment, marked as answer]
There is an inconsistency in your code. For `pressure_adc` you take 3 bytes from the buffer, while using ustruct() only the first two bytes are packed into the `pressure_value`, no matter how long the buffer is, which you supply. You could as well supply self.buf directly. Struct does not know 3 byte quantities. You could use `>i` with a 4 byte buffer, where the first byte is set to 0 and the remaining three bytes are from your returned data.
Edit: like
`pressure_value=ustruct.unpack(">i", b'\x00' + self.buf)[0]`

**Reply by labodezao** [reply]
what is the difference between l and i in struct ? I don(t understand the docs

**Reply by robert-hh** [reply]
> what is the difference between l and i in struct ?

None. Both refer to a 4 byte signed integer.

**Reply by jimmo** [reply]
> what is the difference between l and i in struct ? I don(t understand the docs
> None. Both refer to a 4 byte signed integer.

@labodezao @robert-hh See https://docs.python.org/3/library/struct.html#format-characters

`i` is signed, and `I` is unsigned.  (Same for all integer types, upper case is unsigned).

**Reply by robert-hh** [reply]
@jimmo This seems to be the old problem with the Grotesque fonts like Helvetica and Arial: The capital `i` and the lower case `l`  look the same. I had seen the `l`, you have seen the `I`.
